accounts-db/src/accounts.rs

```diff
@@ -586,10 +624,14 @@ impl Accounts {
         readonly_keys: Vec<&Pubkey>,
     ) {
         for k in writable_keys {
-            account_locks.unlock_write(k);
+            if account_locks.is_locked_write(k) {
+                account_locks.unlock_write(k);
+            }
         }
         for k in readonly_keys {
-            account_locks.unlock_readonly(k);
+            if account_locks.is_locked_readonly(k) {
+                account_locks.unlock_readonly(k);
```

Comment:
this `if` and the preceding `if` will effectively nullify `debug_assert!()`s inside `unlock_readonly()` and `unlock_write()`. I think this new behavior is a bit lenient.

Reply:
Agree with sentiment, and would extend it to the above write-locks as well: https://github.com/anza-xyz/agave/pull/1624/files#r1651320786

Reply:
right, this makes sense. As @apfitzge suggested, write locks also need to be tracking the number of outstanding write-locks.